lib/kintsugi/apply_change_to_project.rb

```diff
@@ -46,10 +57,117 @@ def apply_change_to_project(project, change)
 
     private
 
+    def apply_main_group_change(project, main_group_change)
+      additions, removals, diffs = classify_group_and_file_changes(main_group_change, "")
+      apply_group_additions(project, additions)
+      apply_file_changes(project, additions, removals)
+      apply_group_and_file_diffs(project, diffs)
+      apply_group_removals(project, removals)
+    end
+
+    def classify_group_and_file_changes(change, path)
+      children_changes = change["children"] || {}
+      removals = flatten_change(children_changes[:removed], path)
+      additions = flatten_change(children_changes[:added], path)
+      diffs = [[change, path]]
+      subchanges_of_change(children_changes).each do |key, subchange|
+        sub_additions, sub_removals, sub_diffs =
+          classify_group_and_file_changes(subchange, join_path(path, key))
+        removals += sub_removals
+        additions += sub_additions
+        diffs += sub_diffs
+      end
+
+      [additions, removals, diffs]
+    end
+
+    def flatten_change(change, path)
+      entries = (change || []).map do |child|
+        [child, path]
+      end
+      group_entries = entries.map do |group, _|
+        next if group["children"].nil?
+
+        flatten_change(group["children"], join_path(path, group["displayName"]))
+      end.compact.flatten(1)
+      entries + group_entries
+    end
+
+    def apply_group_additions(project, additions)
+      additions.each do |change, path|
+        next unless %w[PBXGroup PBXVariantGroup].include?(change["isa"])
+
+        containing_group = path.empty? ? project.main_group : project[path]
+        new_group = project.new(Module.const_get("Xcodeproj::Project::#{change["isa"]}"))
+        containing_group.children << new_group
+        add_attributes_to_component(new_group, change, path, ignore_keys: ["children"])
+      end
+    end
+
+    def apply_file_changes(project, additions, removals)
+      addition_to_paths = additions.to_multi_h
+      removal_to_references = removals.to_multi_h.map do |change, paths|
+        references = paths.map do |containing_path|
+          project[join_path(containing_path, change["displayName"])]
+        end
+
+        [change, references]
+      end.to_h
+
+      additions.each do |change, path|
+        next unless change["isa"] == "PBXFileReference"
+
+        containing_group = path.empty? ? project.main_group : project[path]
+
+        if (removal_to_references[change] || []).compact.empty?
```

Comment:
Why not doing `compact` when getting the references?
Otherwise you need to do compact in the line below: `removal_to_references[change].compact.length`

Reply:
We actually should never compact, since we cannot move files which don't exist, as we no longer have their build file